File: app/Modules/Storefront/Http/Controllers/Panel/ProductController.php
<?php

namespace App\Modules\Storefront\Http\Controllers\Panel;

use App\Modules\Storefront\Http\Controllers\Controller;
use App\Modules\Storefront\Models\ParishNote;
use App\Modules\Storefront\Models\Product;
use App\Modules\Storefront\Services\ShopStatsService;
use Illuminate\Http\Request;
use Illuminate\Support\Str;
use Illuminate\Validation\Rule;
use Inertia\Inertia;

class ProductController extends Controller
{
    /**
     * Lista parafii z filtrem po statusie i wyszukiwarką (nazwa / miasto / województwo).
     */
    public function index(Request $request)
    {
        // Filtr statusu (zakładki) — tylko dozwolone klucze.
        $status = in_array($request->query('status'), array_keys(Product::STATUSES), true)
            ? $request->query('status') : null;
        $q = trim((string) $request->query('q', ''));

        $query = Product::withCount('orders')->orderBy('id');

        if ($status) {
            $query->where('status', $status);
        }
        if ($q !== '') {
            $query->where(function ($sub) use ($q) {
                $sub->where('name', 'like', "%{$q}%")
                    ->orWhere('city', 'like', "%{$q}%")
                    ->orWhere('voivodeship', 'like', "%{$q}%");
            });
        }

        $parishes = $query->get();

        // Liczniki per status (niezależne od aktywnego filtra/wyszukiwania).
        $statusCounts = Product::query()
            ->selectRaw('status, COUNT(*) as c')->groupBy('status')->pluck('c', 'status');
        $total = Product::count();

        // Zakładki-liczniki (filtr statusu), zachowując wyszukiwanie.
        $tabs = [[
            'key' => null, 'label' => 'Wszystkie', 'count' => $total,
            'url' => route('panel.products.index', array_filter(['q' => $q])), 'active' => ! $status,
        ]];
        foreach (Product::STATUSES as $key => $label) {
            $tabs[] = [
                'key' => $key, 'label' => $label, 'count' => (int) ($statusCounts[$key] ?? 0),
                'url' => route('panel.products.index', array_filter(['status' => $key, 'q' => $q])),
                'active' => $status === $key,
            ];
        }

        return Inertia::render('Panel/Products/Index', [
            'items' => $parishes->map(fn (Product $p) => $this->presentRow($p))->values(),
            'tabs' => $tabs,
            'q' => $q,
            'indexUrl' => route('panel.products.index'),
            'createUrl' => route('panel.products.create'),
        ]);
    }

    /** Serializacja wiersza listy parafii dla React. */
    private function presentRow(Product $p): array
    {
        [$bg, $fg] = $p->statusColors();

        return [
            'id' => $p->id,
            'name' => $p->name,
            'city' => $p->city,
            'voivodeship' => $p->voivodeship,
            'status_label' => $p->statusLabel(),
            'status_bg' => $bg,
            'status_fg' => $fg,
            'tag_uid' => $p->tag_uid,
            'orders_count' => $p->orders_count,
            'edit_url' => route('panel.products.edit', $p),
            'stats_url' => route('panel.products.stats', $p),
        ];
    }

    public function create()
    {
        return $this->form(new Product());
    }

    public function store(Request $request)
    {
        $data = $this->validated($request);

        $product = Product::create([
            ...$data,
            'slug' => $this->uniqueSlug($data['name']),
            'main_image' => $this->storeMainImage($request),
        ]);

        $this->storeGallery($request, $product);

        return redirect()->route('panel.products.index')->with('success', 'Produkt dodany.');
    }

    public function edit(Product $product)
    {
        $product->load('notes', 'images');

        return $this->form($product);
    }

    /** Wspólny render formularza dodawania/edycji parafii (Inertia). */
    private function form(Product $product)
    {
        $exists = $product->exists;
        [$bg, $fg] = $product->statusColors();

        return Inertia::render('Panel/Products/Form', [
            'product' => [
                'id' => $product->id,
                'exists' => $exists,
                'name' => $product->name,
                'purpose' => $product->purpose,
                'city' => $product->city,
                'voivodeship' => $product->voivodeship,
                'phone' => $product->phone,
                'website' => $product->website,
                'status' => $product->status ?? 'kontakt',
                'status_label' => $product->statusLabel(),
                'status_bg' => $bg,
                'status_fg' => $fg,
                // Cena w formularzu w złotówkach (kropka dziesiętna), pusta dla nowej.
                'price' => $exists ? number_format($product->price / 100, 2, '.', '') : '',
                'tag_uid' => $product->tag_uid,
                'pickup_instruction' => $product->pickup_instruction,
                'description_html' => $product->description_html,
                'main_image_url' => $product->main_image ? asset('storage/' . $product->main_image) : null,
            ],
            'images' => $exists
                ? $product->images->map(fn ($img) => ['id' => $img->id, 'url' => asset('storage/' . $img->path)])->values()
                : [],
            'notes' => $exists ? $product->notes->map(fn ($n) => $this->presentNote($n))->values() : [],
            'statuses' => collect(Product::STATUSES)->map(fn ($l, $k) => ['value' => $k, 'label' => $l])->values(),
            'voivodeships' => Product::VOIVODESHIPS,
            'noteTypes' => collect(ParishNote::TYPES)->map(fn ($l, $k) => ['value' => $k, 'label' => $l])->values(),
            'urls' => [
                'store' => route('panel.products.store'),
                'update' => $exists ? route('panel.products.update', $product) : null,
                'index' => route('panel.products.index'),
                'editorUpload' => route('panel.products.editor-upload'),
                'status' => $exists ? route('panel.parishes.status', $product) : null,
                // Szablony URL-i (podmiana __ID__ po stronie React).
                'imageDelete' => $exists ? route('panel.products.images.delete', [$product, '__ID__']) : null,
                'notesStore' => $exists ? route('panel.parishes.notes.store', $product) : null,
                'notesDestroy' => $exists ? route('panel.parishes.notes.destroy', [$product, '__ID__']) : null,
            ],
        ]);
    }

    /** Serializacja notatki CRM dla React. */
    private function presentNote(ParishNote $n): array
    {
        return [
            'id' => $n->id,
            'body' => $n->body,
            'type' => $n->type,
            'type_label' => $n->typeLabel(),
            'author' => $n->author,
            'created_at' => $n->created_at?->format('Y-m-d H:i'),
        ];
    }

    /**
     * Szybka zmiana statusu parafii (kontakt → test → wdrożenie → aktywna).
     * Status 'aktywna' publikuje parafię (active=true), pozostałe ją ukrywają.
     */
    public function status(Request $request, Product $product)
    {
        $data = $request->validate([
            'status' => ['required', 'string', 'in:' . implode(',', array_keys(Product::STATUSES))],
        ]);

        $product->update([
            'status' => $data['status'],
            'active' => $data['status'] === 'aktywna',
        ]);

        return back()->with('success', 'Status parafii zmieniony na: ' . $product->statusLabel() . '.');
    }

    /**
     * Dodanie notatki CRM (AJAX) — zwraca utworzoną notatkę jako JSON.
     */
    public function storeNote(Request $request, Product $product)
    {
        $data = $request->validate([
            'body' => ['required', 'string', 'max:5000'],
            'type' => ['required', 'string', 'in:' . implode(',', array_keys(ParishNote::TYPES))],
        ], [], ['body' => 'treść', 'type' => 'typ']);

        $note = $product->notes()->create([
            'body' => $data['body'],
            'type' => $data['type'],
            'author' => optional($request->user())->name ?? optional($request->user())->email,
            // created_at ma DB default useCurrent(), ale ustawiamy jawnie,
            // by od razu zwrócić datę w odpowiedzi JSON (bez refetcha).
            'created_at' => now(),
        ]);

        return response()->json([
            'id' => $note->id,
            'body' => $note->body,
            'type' => $note->type,
            'type_label' => $note->typeLabel(),
            'author' => $note->author,
            'created_at' => $note->created_at?->format('Y-m-d H:i'),
        ], 201);
    }

    /**
     * Usunięcie notatki CRM (AJAX).
     */
    public function destroyNote(Product $product, ParishNote $note)
    {
        abort_unless($note->product_id === $product->id, 404);

        $note->delete();

        return response()->json(['deleted' => true]);
    }

    public function update(Request $request, Product $product)
    {
        $data = $this->validated($request, $product);

        if ($main = $this->storeMainImage($request)) {
            $data['main_image'] = $main;
        }

        $product->update($data);
        $this->storeGallery($request, $product);

        return redirect()->route('panel.products.index')->with('success', 'Produkt zapisany.');
    }

    public function toggle(Product $product)
    {
        $product->update(['active' => ! $product->active]);

        return back()->with('success', $product->active ? 'Produkt aktywowany.' : 'Produkt dezaktywowany.');
    }

    public function deleteImage(Product $product, int $imageId)
    {
        $product->images()->where('id', $imageId)->delete();

        return back()->with('success', 'Zdjęcie usunięte.');
    }

    public function stats(Product $product, ShopStatsService $stats)
    {
        $total = $stats->summary(productId: $product->id);
        $last30 = $stats->summary(productId: $product->id, days: 30);
        $series = $stats->dailyPurchases(productId: $product->id, days: 30);

        // Przychód formatujemy tu (grosze → PLN), reszta metryk to liczby całkowite.
        $withRevenue = fn (array $s) => [...$s, 'revenue' => ShopStatsService::formatPln($s['revenue'])];

        return Inertia::render('Panel/Products/Stats', [
            'product' => ['name' => $product->name],
            'total' => $withRevenue($total),
            'last30' => $withRevenue($last30),
            // Lejek sprzedażowy (łącznie) — wartości bezwzględne, skalowanie w React.
            'funnel' => [
                ['label' => 'Otwarcia tagów', 'value' => $total['opens']],
                ['label' => 'Kliki „Kup"', 'value' => $total['clicks']],
                ['label' => 'Zakupy', 'value' => $total['purchases']],
            ],
            'series' => $series,
            'indexUrl' => route('panel.products.index'),
        ]);
    }

    /**
     * Upload zdjęcia z edytora WYSIWYG — zwraca URL do wstawienia w treść.
     */
    public function uploadEditorImage(Request $request)
    {
        $request->validate(['image' => ['required', 'image', 'max:8192']]);

        $path = $request->file('image')->store('products/editor', 'public');

        return response()->json(['url' => asset('storage/' . $path)]);
    }

    private function validated(Request $request, ?Product $product = null): array
    {
        $data = $request->validate([
            'name' => ['required', 'string', 'max:255'],
            'purpose' => ['nullable', 'string', 'max:255'],
            'city' => ['nullable', 'string', 'max:255'],
            'price' => ['required', 'regex:/^\d{1,5}([.,]\d{1,2})?$/'],
            'tag_uid' => ['required', 'string', 'max:255', Rule::unique('products', 'tag_uid')->ignore($product?->id)],
            'pickup_instruction' => ['nullable', 'string', 'max:2000'],
            'description_html' => ['nullable', 'string'],
            // Pola CRM:
            'phone' => ['nullable', 'string', 'max:255'],
            'website' => ['nullable', 'string', 'max:255'],
            'voivodeship' => ['nullable', 'string', 'max:255'],
            'status' => ['required', 'string', 'in:' . implode(',', array_keys(Product::STATUSES))],
        ], [], [
            'name' => 'nazwa', 'purpose' => 'cel zbiórki', 'price' => 'cena', 'tag_uid' => 'UID taga',
            'pickup_instruction' => 'instrukcja odbioru', 'description_html' => 'opis',
            'phone' => 'telefon', 'website' => 'strona www', 'voivodeship' => 'województwo',
            'status' => 'status',
        ]);

        // Cena wpisywana w złotówkach — w bazie trzymamy grosze.
        $data['price'] = (int) round(((float) str_replace(',', '.', (string) $data['price'])) * 100);
        // Status steruje publikacją: aktywna => publiczna, pozostałe => lead (ukryta).
        $data['active'] = $data['status'] === 'aktywna';

        return $data;
    }

    private function uniqueSlug(string $name): string
    {
        $base = Str::slug($name);
        $slug = $base;
        $i = 2;
        while (Product::where('slug', $slug)->exists()) {
            $slug = $base . '-' . $i++;
        }

        return $slug;
    }

    private function storeMainImage(Request $request): ?string
    {
        if (! $request->hasFile('main_image')) {
            return null;
        }

        $request->validate(['main_image' => ['image', 'max:8192']]);

        return $request->file('main_image')->store('products', 'public');
    }

    private function storeGallery(Request $request, Product $product): void
    {
        if (! $request->hasFile('gallery')) {
            return;
        }

        $request->validate(['gallery.*' => ['image', 'max:8192']]);

        $sort = (int) $product->images()->max('sort') + 1;
        foreach ($request->file('gallery') as $file) {
            $product->images()->create([
                'path' => $file->store('products', 'public'),
                'sort' => $sort++,
            ]);
        }
    }
}

File: pay-backend/storefront-api/src/main/kotlin/com/supportme/pay/storefront/api/panel/product/ProductController.kt
package com.supportme.pay.storefront.api.panel.product

import com.supportme.pay.storefront.api.storage.FileStorageService
import com.supportme.pay.storefront.domain.entity.ParishNote
import com.supportme.pay.storefront.domain.entity.ParishNoteType
import com.supportme.pay.storefront.domain.entity.Product
import com.supportme.pay.storefront.domain.entity.ProductImage
import com.supportme.pay.storefront.domain.entity.ProductStatus
import com.supportme.pay.storefront.domain.repository.ParishNoteRepository
import com.supportme.pay.storefront.domain.repository.ProductImageRepository
import com.supportme.pay.storefront.domain.repository.ProductRepository
import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.Pattern
import jakarta.validation.constraints.Size
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.multipart.MultipartFile

data class ProductPanelSummary(val id: Long, val name: String, val city: String?, val slug: String, val status: String, val active: Boolean)
data class ProductRequest(
    @field:NotBlank @field:Size(max = 255) val name: String,
    @field:Size(max = 255) val city: String? = null,
    val purpose: String? = null,
    val descriptionHtml: String? = null,
    @field:Size(max = 2000) val pickupInstruction: String? = null,
    /** String, jak w PHP (`regex:/^\d{1,5}([.,]\d{1,2})?$/`) — pozwala na grosze (np. "19,99"), nie tylko całe złote. */
    @field:NotBlank @field:Pattern(regexp = "\\d{1,5}([.,]\\d{1,2})?") val price: String,
    @field:NotBlank @field:Size(max = 255) val tagUid: String,
    @field:Size(max = 255) val phone: String? = null,
    @field:Size(max = 255) val website: String? = null,
    @field:Size(max = 255) val voivodeship: String? = null,
    @field:NotBlank val status: String,
)
data class ProductStatsSummaryResponse(val opens: Long, val views: Long, val clicks: Long, val purchases: Long, val revenuePln: String, val conversionPercent: Double)
data class FunnelPoint(val label: String, val value: Long)
data class ProductStatsResponse(
    val total: ProductStatsSummaryResponse,
    val last30: ProductStatsSummaryResponse,
    val funnel: List<FunnelPoint>,
    val series: List<DailyPurchasePoint>,
)
data class NoteRequest(@field:NotBlank @field:Size(max = 5000) val body: String, val type: String = "kontakt", val author: String? = null)
data class NoteSummary(val id: Long, val body: String, val type: String, val author: String?, val createdAt: String?)

/** Odpowiednik `Panel\ProductController` — CRM parafii/Taca. `status` steruje `active` (publish gate). */
@RestController
@RequestMapping("/api/storefront/panel/products")
class ProductController(
    private val productRepository: ProductRepository,
    private val productImageRepository: ProductImageRepository,
    private val parishNoteRepository: ParishNoteRepository,
    private val fileStorageService: FileStorageService,
    private val statsService: StorefrontStatsService,
) {

    @GetMapping
    fun index(@RequestParam(required = false) status: String?, @RequestParam(required = false) search: String?): List<ProductPanelSummary> {
        var products = productRepository.findAll()
        status?.let { s -> ProductStatus.entries.firstOrNull { it.dbValue == s }?.let { st -> products = products.filter { it.status == st } } }
        search?.takeIf { it.isNotBlank() }?.let { q ->
            val needle = q.lowercase()
            products = products.filter { it.name.lowercase().contains(needle) || it.city?.lowercase()?.contains(needle) == true || it.voivodeship?.lowercase()?.contains(needle) == true }
        }
        return products.map { ProductPanelSummary(it.id!!, it.name, it.city, it.slug, it.status.dbValue, it.active) }
    }

    @PostMapping
    fun store(@Valid @RequestBody body: ProductRequest): ResponseEntity<Any> {
        val status = ProductStatus.entries.firstOrNull { it.dbValue == body.status }
            ?: return ResponseEntity.status(HttpStatus.UNPROCESSABLE_ENTITY).body(mapOf("error" to "Nieprawidłowy status."))
        if (productRepository.findByTagUid(body.tagUid) != null) {
            return ResponseEntity.status(HttpStatus.UNPROCESSABLE_ENTITY).body(mapOf("error" to "Tag NFC już przypisany do innej parafii."))
        }
        val product = productRepository.save(
            Product(
                name = body.name, city = body.city, purpose = body.purpose, slug = generateUniqueSlug(body.name),
                descriptionHtml = body.descriptionHtml, pickupInstruction = body.pickupInstruction, price = parsePriceGrosze(body.price),
                tagUid = body.tagUid, phone = body.phone, website = body.website, voivodeship = body.voivodeship,
                status = status, active = status == ProductStatus.AKTYWNA,
            ),
        )
        return ResponseEntity.status(HttpStatus.CREATED).body(mapOf("id" to product.id!!))
    }

    @PutMapping("/{id}")
    fun update(@PathVariable id: Long, @Valid @RequestBody body: ProductRequest): ResponseEntity<Any> {
        val product = productRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        val status = ProductStatus.entries.firstOrNull { it.dbValue == body.status }
            ?: return ResponseEntity.status(HttpStatus.UNPROCESSABLE_ENTITY).body(mapOf("error" to "Nieprawidłowy status."))
        if (body.tagUid != product.tagUid && productRepository.findByTagUid(body.tagUid) != null) {
            return ResponseEntity.status(HttpStatus.UNPROCESSABLE_ENTITY).body(mapOf("error" to "Tag NFC już przypisany do innej parafii."))
        }

        product.name = body.name
        product.city = body.city
        product.purpose = body.purpose
        product.descriptionHtml = body.descriptionHtml
        product.pickupInstruction = body.pickupInstruction
        product.price = parsePriceGrosze(body.price)
        product.tagUid = body.tagUid
        product.phone = body.phone
        product.website = body.website
        product.voivodeship = body.voivodeship
        // Status steruje publikacją: aktywna => publiczna, pozostałe => lead (ukryta) — jak w PHP.
        product.status = status
        product.active = status == ProductStatus.AKTYWNA
        productRepository.save(product)
        return ResponseEntity.ok(mapOf("id" to product.id!!))
    }

    /** Jak `(int) round(((float) str_replace(',', '.', $price)) * 100)` w PHP. */
    private fun parsePriceGrosze(price: String): Int = Math.round(price.replace(',', '.').toDouble() * 100).toInt()

    @PostMapping("/{id}/main-image")
    fun uploadMainImage(@PathVariable id: Long, @RequestParam file: MultipartFile): ResponseEntity<Any> {
        val product = productRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        product.mainImage?.let { fileStorageService.deletePublic(it) }
        product.mainImage = fileStorageService.storePublic(file, "products")
        productRepository.save(product)
        return ResponseEntity.ok(mapOf("mainImage" to product.mainImage!!))
    }

    @PostMapping("/{id}/gallery")
    fun uploadGalleryImages(@PathVariable id: Long, @RequestParam files: List<MultipartFile>): ResponseEntity<Any> {
        val product = productRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        val maxSort = productImageRepository.findAllByProductOrderBySortAsc(product).maxOfOrNull { it.sort } ?: -1
        files.forEachIndexed { index, file ->
            val path = fileStorageService.storePublic(file, "products")
            productImageRepository.save(ProductImage(product = product, path = path, sort = maxSort + 1 + index))
        }
        return ResponseEntity.status(HttpStatus.CREATED).body(mapOf("status" to "ok"))
    }

    @DeleteMapping("/{id}/images/{imageId}")
    fun deleteImage(@PathVariable id: Long, @PathVariable imageId: Long): ResponseEntity<Any> {
        // Scope po product_id — jak `$product->images()->where('id',$imageId)` w PHP,
        // inaczej DELETE ...{A}/images/{imageIdNależącyDoB} usuwałby zdjęcie produktu B.
        val image = productImageRepository.findById(imageId).orElse(null)?.takeIf { it.product.id == id }
            ?: return ResponseEntity.notFound().build()
        fileStorageService.deletePublic(image.path)
        productImageRepository.delete(image)
        return ResponseEntity.noContent().build()
    }

    /** Zmiana statusu CRM — `active` przełącza się TYLKO przy statusie `aktywna` (publish gate). */
    @PostMapping("/{id}/status")
    fun updateStatus(@PathVariable id: Long, @RequestBody body: Map<String, String>): ResponseEntity<Any> {
        val product = productRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        val status = ProductStatus.entries.firstOrNull { it.dbValue == body["status"] }
            ?: return ResponseEntity.status(HttpStatus.UNPROCESSABLE_ENTITY).build()
        product.status = status
        product.active = status == ProductStatus.AKTYWNA
        productRepository.save(product)
        return ResponseEntity.ok(mapOf("status" to status.dbValue, "active" to product.active))
    }

    @GetMapping("/{id}/stats")
    fun stats(@PathVariable id: Long): ResponseEntity<ProductStatsResponse> {
        productRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()

        val total = statsService.summary(id, null)
        val last30 = statsService.summary(id, 30)
        val series = statsService.dailyPurchases(id, 30)

        return ResponseEntity.ok(
            ProductStatsResponse(
                total = total.toResponse(),
                last30 = last30.toResponse(),
                funnel = listOf(
                    FunnelPoint("Otwarcia tagów", total.opens),
                    FunnelPoint("Kliki „Kup”", total.clicks),
                    FunnelPoint("Zakupy", total.purchases),
                ),
                series = series,
            ),
        )
    }

    private fun ShopStatsSummary.toResponse() = ProductStatsSummaryResponse(opens, views, clicks, purchases, revenuePln, conversionPercent)

    @GetMapping("/{id}/notes")
    fun notes(@PathVariable id: Long): ResponseEntity<List<NoteSummary>> {
        val product = productRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        return ResponseEntity.ok(parishNoteRepository.findAllByProductOrderByIdDesc(product).map { NoteSummary(it.id!!, it.body, it.type.dbValue, it.author, it.createdAt?.toString()) })
    }

    @PostMapping("/{id}/notes")
    fun storeNote(@PathVariable id: Long, @Valid @RequestBody body: NoteRequest): ResponseEntity<Any> {
        val product = productRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        val type = ParishNoteType.entries.firstOrNull { it.dbValue == body.type }
            ?: return ResponseEntity.status(HttpStatus.UNPROCESSABLE_ENTITY).body(mapOf("error" to "Nieprawidłowy typ notatki."))
        val note = parishNoteRepository.save(ParishNote(product = product, body = body.body, type = type, author = body.author))
        return ResponseEntity.status(HttpStatus.CREATED).body(mapOf("id" to note.id!!))
    }

    @DeleteMapping("/{id}/notes/{noteId}")
    fun destroyNote(@PathVariable id: Long, @PathVariable noteId: Long): ResponseEntity<Any> {
        // Scope po product_id — jak `abort_unless($note->product_id === $product->id, 404)` w PHP.
        val note = parishNoteRepository.findById(noteId).orElse(null)?.takeIf { it.product.id == id }
            ?: return ResponseEntity.notFound().build()
        parishNoteRepository.delete(note)
        return ResponseEntity.noContent().build()
    }

    /** Generyczny upload obrazka edytora WYSIWYG — współdzielony przez BeneficiaryNode i tu. */
    @PostMapping("/upload-editor-image")
    fun uploadEditorImage(@RequestParam file: MultipartFile): ResponseEntity<Any> {
        val path = fileStorageService.storePublic(file, "products/editor")
        return ResponseEntity.ok(mapOf("path" to path))
    }

    private fun generateUniqueSlug(name: String): String {
        val base = name.lowercase().replace(Regex("[^a-z0-9]+"), "-").trim('-').ifBlank { "parafia" }
        var candidate = base
        var i = 2
        while (productRepository.findBySlug(candidate) != null) candidate = "$base-${i++}"
        return candidate
    }
}
